### PR TITLE
feat(manager): :sparkles: add region to openStackMachineSpec.ProviderID field from crd identityRef

### LINKED ISSUE #2183: Multi openstack cloud support with OCCM
/kind feature

**Glossary**

 - **cloud**: I will use "cloud" term to identify an independent openstack cluster, 2 different clouds means 2 different openstack clusters with differents endpoints, credentials, regoin name, keystore ...

**Description**

I would like to be able to create k8s cluster spread across different openstack clusters. I have already underlying network based on VPN tunnels and network interconnections, ready to VMs communications between different clouds.

I am able to create VM in my different clouds with the help of `spec.template.spec.identityRef` from `OpenStackMachineTemplate` which permit to override `idetityRef` define in parent `OpenStackCluster` object.

It works well ! My issue concern integration with OCCM (Openstack Cloud Controller Manager).

The only way yo manage multiple clouds with OCCM seems to be to deploy 1 OCCM instance per cloud by setting env variables `OS_CCM_REGIONAL="true"` and `OS_V1_INSTANCES="true"`.

In this way each OCCM manage his own cloud by adding OS region name (a limitation could be that 2 differents clouds should have different regions names but I don't think that is really an issue). The "feature flag" implementation is [here](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/openstack/instances.go#L371).
OCCM with this env variables manage VMs with providerId format like `openstack://region_name/uuid`
k8s nodes `spec.providerID` is immutable and created by kubeadm in the usecase of CAPO, so I configure my VMs with following configuration :
```yaml
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
metadata:
  name: pool-region-one
  annotations:
    controlplane.cluster.x-k8s.io/skip-coredns: ""
    controlplane.cluster.x-k8s.io/skip-kube-proxy: ""
spec:
  template:
    spec:
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cloud-provider: external
            provider-id: openstack://region-one/'{{ instance_id }}'
...
```
In that way k8s node are properly provision with the right providerID anc correctly recognize by the right OCCM.

**Missing feature**

The issue is now on CAPO implementation which hardcode providerID in format `openstack:///uuid` in his machine crd object with no possibility to configure endoint field with region name
```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Machine
spec:
  providerID: openstack:///uuid
  ...
```
I explain this scenario in an issue concerning usecase without kubeadm and unset providerID with usage of multiple OCCM in same k8s cluster https://github.com/kubernetes/cloud-provider-openstack/pull/2590#issuecomment-2376822814 , 
**TLDR**: CAPI see k8s node with providerID `openstack://region_name/uuid` but CAPO see machine with providerID `openstack:///uuid` (with same uuid) and it is stuck in this state.

**Code to patch**

I think that CAPO machine crd providerID is hardcoded [here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/controllers/openstackmachine_controller.go#L425) but I don't find proper way to get region name here, to template this instruction.

**How to patch it ?**

- I can retreive identityRef fields so technically I can read k8s secret and parse OS config to get region name from configuration.
- Add an optional field in [OpenStackMachineSpec](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1beta1/openstackmachine_types.go#L95) "region_name" which will be added in providerID if it is fill.
- I didn't find better way yet... Maybe @mdbooth or someone could give me a good hint :crossed_fingers:

### PR DESCRIPTION
**Pull request by MatthieuFin**
**What this PR does / why we need it**:

To made CAPO compatible with OCCM multi-cloud deployment we need to manage providerID format with region code as introduced [here](https://github.com/kubernetes/cloud-provider-openstack/issues/1900) (`openstack://region/instance_uuid`). Use annotation from OpenstackMachine to define region in providerID field.

An empty or undefined annotation permit to be fully backward compatible. Moreover use annotation permit to manage some cluster with new OCCM provider-id string and other cluster with old format by same CAPO instance.

Fixes #2183

**Special notes for your reviewer**:

- As discussed in issue #2183 I first try to implement k8s secret call to get region name from config [here](https://github.com/MatthieuFin/cluster-api-provider-openstack/commit/04a4e609617d51249142d3328b3c017ed3d9c72a), it works well but it's not backward compatible. And if I setup a feature flag wiath a cli arg I loose the possibility to manage some k8s cluster with new providerID format beside of some k8s cluster with old providerID format managed by same CAPO instance.
- This feature need https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2191 merged to work correctly !

### CODE REVIEW

File: controllers/openstackmachine_controller.go [hunk 1]
```diff
@@ -422,7 +422,7 @@ func (r *OpenStackMachineReconciler) reconcileMachineState(scope *scope.WithLogg
 		conditions.MarkTrue(openStackMachine, infrav1.InstanceReadyCondition)
 
 		// Set properties required by CAPI machine controller
-		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack:///%s", *openStackServer.Status.InstanceID))
+		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack://%s/%s", openStackMachine.ObjectMeta.Annotations["cluster.x-k8s.io/openstack.cloud"], *openStackServer.Status.InstanceID))
```
Comment (mdbooth) [inline controllers/openstackmachine_controller.go:425]: As [discussed in slack](https://kubernetes.slack.com/archives/CFKJB65G9/p1728981106884829), lets look into adding this to the API rather than as an annotation. This should be fine as long as any new field is optional, and the behaviour remains unchanged when it is not specified.

Something to think about: we permit the modification of identityRef on objects, but if we added `region` and it became part of the `providerID`, should it be immutable?

Something else to note here is that we need to preserve the immutability of `providerID`. This isn't a problem in the current code, as if we have an instance it's always going to have the same ID. With this change this is no longer true, as the behaviour can change across an upgrade. However we do this, I suggest that we add a check here and only set `ProviderID` if it is currently unset.

**Comment by mdbooth** [comment]
I've spent much of the morning thinking this through and discussing with @stephenfin and Sean Mooney. Based on that, I'd like to define the following contract:

* The format of `providerID` is openstack://`region`/`server uuid`
* If region is defined, it specifically means that the server with `server uuid` was created by and is known to the Nova API in `region`.
* It explicitly does not make any direct assertions about any other OpenStack resources. It refers specifically to the Nova server.
* region may be empty, in which case consumers of providerID which require a region must have some other way to obtain it. The lack of a region cannot be used to make assumptions about region.

It's also worth restating that providerID is immutable, so any change we make to it will not be applied to existing Machines.

In case anybody else was also fuzzy about the details, here's a restatement of my understanding of OpenStack regions: regions are effectively only a key used in the catalog of services returned by keystone during authentication. Region is not used during authentication. Keystone returns all service endpoints, and the client picks one from the desired region.

This means it makes no sense to use a region returned by keystone. Instead we should record the region we used to select and endpoint. i.e. We use the *input* from the API, not the *output* from keystone.

I propose we could add `Region` as an optional, immutable value in `OpenStackIdentityReference`, perhaps:
```go
	// Region specifies an OpenStack region to use. If specified, it overrides
	// any value in clouds.yaml. If specified for an OpenStackMachine, it's
	// value will be included in providerID.
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="region is immutable"
	// +optional
	Region *string `json:"region,omitempty"`
```

I would expect us to use that value explicitly here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/934d61d38a26210b00811c9a3d354e92f32d09f7/pkg/scope/provider.go#L182-L188

as well as when setting `providerID`.

**Comment by MatthieuFin** [comment]
Personally, I prefer don't use the new field `Region` from `OpenStackIdentityReference` in OS client creation https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/934d61d38a26210b00811c9a3d354e92f32d09f7/pkg/scope/provider.go#L182-L188 

If you have 2 providers which declare same region name, in this case you cannot differentiate these regions. If we keep them manage separately you could have `openstack://provider1-region1/instance_uuid` and `openstack://provider2-region1/instance_uuid`. and keep region field in client configuration to `region1` in both cases.

Currently OCCM don't allow this but maybe a day it will be supported.

**Comment by MatthieuFin** [comment]
Ok, that’s good for me, let’s see if we need to explore this topic further later! And you're right, we'll probably never need to replace the region name, at least it's too early to worry about that.

File: api/v1beta1/identity_types.go [hunk 2]
```diff
@@ -28,6 +28,13 @@ type OpenStackIdentityReference struct {
 	// CloudName specifies the name of the entry in the clouds.yaml file to use.
 	// +kubebuilder:validation:Required
 	CloudName string `json:"cloudName"`
+
+	// Region specifies an OpenStack region to use. If specified, it overrides
+	// any value in clouds.yaml. If specified for an OpenStackMachine, its
+	// value will be included in providerID.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="region is immutable"
```
Comment (mdbooth) [inline api/v1beta1/identity_types.go:35]: Please can you add a test to https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/test/e2e/suites/apivalidations/openstackmachine_test.go to validate that the field is indeed immutable. I think we only need to test it for a single object, and OpenStackMachine makes sense. We should test:
* The value can be set on creation
* The value cannot be modified after creation
* The value cannot be unset after creation

File: controllers/openstackmachine_controller.go [hunk 3]
```diff
@@ -422,7 +422,7 @@ func (r *OpenStackMachineReconciler) reconcileMachineState(scope *scope.WithLogg
 		conditions.MarkTrue(openStackMachine, infrav1.InstanceReadyCondition)
 
 		// Set properties required by CAPI machine controller
-		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack:///%s", *openStackServer.Status.InstanceID))
+		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack://%s/%s", openStackMachine.Spec.IdentityRef.Region, *openStackServer.Status.InstanceID))
```
Comment (mdbooth) [inline controllers/openstackmachine_controller.go:425]: How can we get some test coverage of this?

**Review by mdbooth — COMMENTED** [review]
This looks great, and thanks for getting this done so quickly. I'm +1 on everything here.

As mentioned we should have apivalidations coverage of all API validations. They're very easy to write and there are lots of examples to copy from.

Ideally I'd also like some e2e test coverage of this. I think we can add this fairly easily by doing some like modifying the MachineDeployment template of the workers of one of the existing e2e scenarios and adding validation in the test suite that providerID is set on the CAPI Machine the way we expect.

Ping me on Slack if I can help with this.

File: controllers/openstackmachine_controller.go (hunk 3 above)
Comment (mdbooth) [reply controllers/openstackmachine_controller.go:425]: Whoops! IdentityRef is a pointer and is usually nil.

Incidentally, wouldn't these errors be so much simpler to see if you'd had to write something like `openStackMachine.Spec.IdentityRef->Region` instead? Go automatic dereferencing is insidious.

File: controllers/openstackmachine_controller.go [hunk 4]
```diff
@@ -422,7 +422,7 @@ func (r *OpenStackMachineReconciler) reconcileMachineState(scope *scope.WithLogg
 		conditions.MarkTrue(openStackMachine, infrav1.InstanceReadyCondition)
 
 		// Set properties required by CAPI machine controller
-		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack://%s/%s", openStackMachine.Spec.IdentityRef.Region, *openStackServer.Status.InstanceID))
+		openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack://%s/%s", (*openStackMachine.Spec.IdentityRef).Region, *openStackServer.Status.InstanceID))
```
Comment (mdbooth) [inline controllers/openstackmachine_controller.go:425]: That's still going to panic when IdentityRef is nil. How about something like:

```go
var region string
if openStackMachine.Spec.IdentityRef != nil {
    region = openStackMachine.Spec.IdentityRef.Region
}
openStackMachine.Spec.ProviderID = ptr.To(fmt.Sprintf("openstack://%s/%s", region, *openStackServer.Status.InstanceID))
```

Comment (MatthieuFin) [reply controllers/openstackmachine_controller.go:425]: Oh, IdentityRef could be nil I did'nt get It !

I'll modify that Thx.

File: api/v1beta1/identity_types.go [hunk 5]
```diff
@@ -18,6 +18,7 @@ package v1beta1
 
 // OpenStackIdentityReference is a reference to an infrastructure
 // provider identity to be used to provision cluster resources.
+// +kubebuilder:validation:XValidation:rule="(!has(self.region) && !has(oldSelf.region)) || self.region == oldSelf.region",message="region is immutable"
```
[the remark this answers is not in the document]
Comment (MatthieuFin) [reply api/v1beta1/identity_types.go:21]: yes, `"self == oldSelf"` as you proposed, permit to set a value on an existing machine without region field defined. To do that I had to create rule on top level object because `has(self)` seems not valid syntax.

**Comment by mdbooth** [comment]
This looks great. Can you squash the commits (into a single commit with a single, coherent description) and I'll approve. Thanks!

**Comment by mdbooth** [comment]
`providerID` looks normal on unaffected machines, e.g. https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/2193/pull-cluster-api-provider-openstack-e2e-test/1846816338392125440/artifacts/clusters/bootstrap/resources/e2e-tk3rp9/Machine/cluster-e2e-tk3rp9-control-plane-2p6jt.yaml

```yaml
providerID: openstack:///d8a589fa-98ca-4999-92da-bf8d3a2678d3
```

As mentioned we still need some test coverage of this actually working. I'll approve on the basis that:
* it's a new feature, so it has no current users to regress
* our existing test coverage shows it does not regress users who aren't using it
* the API looks good
* manual testing by @MatthieuFin

If we want to continue working in the future, though, we'll need a test.

**Comment by MatthieuFin** [comment]
I'm just currently testing on my side before squash that.

I did'nt write e2e tests but I agree we need it to be future proof.

I'm ok to try writing some tests about that, I'll probably ping you on slack for help on that purpose.

File: api/v1beta1/identity_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdbooth) [reply api/v1beta1/identity_types.go:21]: Thanks. Turns out I was missing some tests in the image API and this was broken: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2197

[the remark this answers is not in the document]
Comment (MatthieuFin) [reply api/v1beta1/identity_types.go:21]: Glad to have been helpful !
